### ISSUE TITLE
Add a more practical way to listen on the relay protocol

### ISSUE THREAD

**Issue by tomaka** [libp2p/rust-libp2p#390]
Let's say you're listening for the `floodsub` protocol, and you want to add support for relays.
If you just combine connection upgrades, you will obtain an enum that is either `floodsub`, an intermediate relay, or a relay destination point.
However when you obtain a relay destination point, you want to negotiate floodsub on top of it, and it's not possible to do so right now.

Similarly it is currently impossible to accept either "`floodsub` over `secio`" or just `floodsub`.

We should have some sort of `LoopTransport` that is created from a `Transport`, a `ConnectionUpgrade` that represents the intermediate layer, and a `ConnectionUpgrade` that represents the final layer.

The exact details are probably not easy though.

**Comment by tomaka** [comment]
No longer relevant after #565
